### PR TITLE
Integrate effluent calculations

### LINKED ISSUE #1376: [Feed] Populate logic in `RUFAS/routines/feed_storage/sileage.py`
Populate the logics of the following methods in `RUFAS/routines/feed_storage/sileage.py`:

- [ ] `calculate_protein_loss`
- [ ] `calculate_dry_matter_loss_to_gas`
- [ ] `calculate_dry_matter_loss_to_effluent`

[Design doc is here](https://docs.google.com/document/d/1C86lch2dkYBERR-qLDmE7L7n1gYQjETrIKUyrQpWrPs/edit)

* For questions, reach out to SME: @KevinP-B and/or SWE @PooyaHekmati . Their reviews are needed for any linked PRs.
* Size the PRs reasonably, not all the methods have to be implemented in a single PR.
* docstrings, function signatures, and unit tests need to be updated accordingly.
* I suggest reading and understanding the design doc before attempting to write the code.

### PR DESCRIPTION
**Pull request by Isgotkowitz**
This PR ties the helper routines in `Silage` together so that effluent loss is fully calculated.

### Context

Issue(s) closed by this pull request: closes #1376

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

The method `process_degradations` is implemented in `Silage`, which calculates and processes losses of dry matter and moisture to effluent. It also calls the `process_degradations` in `Storage`, ensuring that both effluent and fermentative loss is processed for ensiled crops.

The subroutines for calculating the fractional loss coefficients of crude protein and non-protein nitrogen are also in this PR.

The method `reset_mass_attributes_after_loss` is updated to also process loss of moisture, in addition to dry matter mass.

### Why

Loss in ensiled crops happens through two processes: fermentation and effluence. Fermentation has already been implemented in `Storage` (this process is shared with `Baleage`), so effluent loss processing needs to be implemented in `Silage`. The code is organized so that `Silage` loses mass and nutrients to both effluence and fermentation while reusing as much code as possible.

The helper routines calculate fractional loss coefficients, which determine the quantity of nutrient loss when effluence occurs. These are dynamic for effluent loss (unlike for fermentation and hay loss) so they need methods to calculate them.

The routine for updating the mass attributes is changed because effluence is a loss of both dry matter and moisture, so both need to be accounted for when recalculating a crop's mass attributes.

### Test plan

Changes have been tested by updating the unit tests and ensuring that they pass. Special attention has been placed on the test cases for calculating the non-protein nitrogen and crude protein loss coefficients - the expected values seem much higher than other fractional loss coefficients, and I suspect there is still an edge case that is not protected against but I don't know how likely it is that this will ever come up in an actual simulation.

### Other
Earlier in the review process it was identified that the equations were incorrectly documented, so the documentation for equations for recalculating crude protein and non-protein nitrogen were updated in this PR as well.

### CODE REVIEW

File: tests/test_feed_storage_module/test_silage.py [hunk 1]
```diff
@@ -84,6 +138,29 @@ def test_calculate_moisture_loss_to_effluent(silage: Silage, max_effluent: float
...
+) -> None:
+    """Tests calculate_non_protein_nitrogen_loss_coefficient in Silage."""
+    actual = silage.calculate_non_protein_nitrogen_loss_coefficient(npn, cp, loss_frac)
+
+    assert pytest.approx(actual) == expected
+
+
+@pytest.mark.parametrize(
+    "cp,loss_frac,expected", [(5.6, 0.033, 5.78086866), (2.2, 0.04, 2.27916666), (0.0, 0.05, -0.015789473)]
+)
```
Comment (Isgotkowitz) [inline tests/test_feed_storage_module/test_silage.py]: @KevinP-B I think these fractional loss coefficients (the "expected" values) are off - they seem much higher than the fractional loss coefficients for hay and fermentation, and I don't think they should ever be negative.

Comment (KevinP-B) [reply tests/test_feed_storage_module/test_silage.py]: There are cases where the fraction of a nutrient increases during storage, e.g. if DM is lost but none of the lost is the nutrient of interest, the relative proportion of the nutrient increases. However, I'm not sure that's what we're seeing. I will follow up.

**Review by KevinP-B — COMMENTED** [review]
This looks like it addresses the major issues in implementation and equation errors that we discussed.

**Review by KevinP-B — APPROVED** [review]
This looks like it addresses the major issues in implementation and equation errors that we discussed.

File: RUFAS/routines/feed_storage/silage.py [hunk 2]
```diff
@@ -38,21 +43,46 @@ def __init__(self, capacity: float = float("inf")):
...
-            Weather instance containing all weather information for the simulation.
-        time : Time
-            Time instance tracking the current time of the simulation.
-
-        Notes
-        -----
-        This method also records the total amount of gaseous dry matter loss happened from all stored crops.
-
+        Processes the losses of nutrients and mass to effluent in the ensiled crops, calls the parent implementation of
+        of `process_degradations` to handle the fermentative loss.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/silage.py:47]: I would add variable definitions to the docstrings since this method is public (even if it's the same as the parent function).

File: RUFAS/routines/feed_storage/silage.py [hunk 3]
```diff
@@ -127,6 +157,68 @@ def calculate_moisture_loss_to_effluent(self, estimated_maximum_effluent: float,
...
+        .. [1] Feed Storage Scientific Documentation, equation 2.2.1.2
+
+        """
+        npn_fraction = initial_non_protein_nitrogen * GeneralConstants.PERCENTAGE_TO_FRACTION
+        cp_fraction = initial_crude_protein * GeneralConstants.PERCENTAGE_TO_FRACTION
+
+        numerator = npn_fraction * cp_fraction - 0.3 * loss_fraction
+        denominator = cp_fraction - 0.3 * loss_fraction
+
+        new_npn_fraction = numerator / denominator
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/silage.py:202]: Error handling in case of a ZeroDivisionError necessary?

File: RUFAS/routines/feed_storage/silage.py [hunk 4]
```diff
@@ -127,6 +157,68 @@ def calculate_moisture_loss_to_effluent(self, estimated_maximum_effluent: float,
...
+        float
+            Percentage of crude protein remaining in the stored crop.
+
+        References
+        ----------
+        .. [1] Feed Storage Scientific Documentation, equation 2.2.1.1
+
+        """
+        new_fraction = (initial_crude_protein * GeneralConstants.PERCENTAGE_TO_FRACTION - 0.3 * loss_fraction) / (
+            1 - loss_fraction
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/silage.py:231]: Error handling in case of a ZeroDivisionError necessary?

File: RUFAS/routines/feed_storage/storage.py [hunk 5]
```diff
@@ -181,7 +181,7 @@ def process_degradations(self, weather: Weather, time: Time) -> None:
             )
 
             crop.last_time_degraded = copy.deepcopy(time)
-            self.reset_mass_attributes_after_loss(crop, gaseous_dry_matter_loss)
+            self.reset_mass_attributes_after_loss(crop, gaseous_dry_matter_loss, 0.0)
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]: Maybe add this as a keyword arg
```suggestion
            self.reset_mass_attributes_after_loss(crop, gaseous_dry_matter_loss, moisture_loss=0.0)
```

File: RUFAS/routines/feed_storage/storage.py [hunk 6]
```diff
@@ -199,21 +199,30 @@ def give_feed(self, amount: float, crop_type: CropType) -> None:
...
-        matter.
+        Resets the dry mass, fresh mass, and dry matter percentage attributes in a stored crop after loss of both dry
+        matter and moisture.
 
         Parameters
         ----------
         crop : HarvestedCrop
             The stored crop that has lost dry matter.
         dry_matter_loss : float
             Amount of dry matter the crop lost on the current day in kg.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
            Amount of dry matter the crop lost on the current day (kg).
```

File: RUFAS/routines/feed_storage/storage.py [hunk 7]
```diff
@@ -199,21 +199,30 @@ def give_feed(self, amount: float, crop_type: CropType) -> None:
...
+        moisture_loss : float
+            Amount of moisture (water) the crop lost on the current day in kg.
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
            Amount of moisture (water) the crop lost on the current day (kg).
```

File: RUFAS/routines/feed_storage/storage.py [hunk 8]
```diff
@@ -199,21 +199,30 @@ def give_feed(self, amount: float, crop_type: CropType) -> None:
...
+
+        Notes
+        -----
+        The amount of dry matter mass remaining is calculating first, then the remaining amount of fresh mass. After
```
Comment (ew3361zh) [inline RUFAS/routines/feed_storage/storage.py]:
```suggestion
        The amount of dry matter mass remaining is calculated first, then the remaining amount of fresh mass. After
```

**Review by ew3361zh — APPROVED** [review]
LGTM. Nice work, Ed and Kevin! Small batch of comments on docstrings.

File: RUFAS/routines/feed_storage/storage.py (hunk 5 above)
Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/storage.py]: Good suggestion

File: RUFAS/routines/feed_storage/storage.py (hunk 8 above)
Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/storage.py]: Nice catch

File: RUFAS/routines/feed_storage/silage.py (hunk 3 above)
Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/silage.py:202]: It's definitely a good idea, thanks Niko

File: RUFAS/routines/feed_storage/silage.py (hunk 4 above)
Comment (Isgotkowitz) [reply RUFAS/routines/feed_storage/silage.py:231]: Same as above
